src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CollectionModelBinder.cs

```diff
@@ -164,12 +161,20 @@ protected object CreateInstance(Type targetType)
                 };
 
                 object boundValue = null;
-                var result =
-                    await bindingContext.OperationBindingContext.ModelBinder.BindModelAsync(innerBindingContext);
-                if (result != null && result.IsModelSet)
+
+                using (bindingContext.PushContext(
+                    elementMetadata,
+                    fieldName: bindingContext.FieldName,
+                    modelName: bindingContext.ModelName,
+                    model: null))
```

Comment:
Pushing/popping guarantees an entirely reset internal state per item, setting Result to null isn't as strong. It was probably outside of the loop before to reduce allocations, which isn't necessary with push/pop structs onto a stack.